### PR TITLE
Add an end-of-input format

### PR DESCRIPTION
**Pull request by brendanzab**

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -234,6 +237,7 @@ impl Format {
         match self {
             Format::Fail => false,
             Format::Empty => true,
+            Format::EndOfInput => false,
```
Comment (brendanzab) [inline src/main.rs:240]: Unsure about this!

File: src/main.rs [hunk 2]
```diff
@@ -303,6 +307,7 @@ impl Switch {
         match f {
             Format::Fail => Ok(()),
             Format::Empty => self.add_next(index, depth, next),
+            Format::EndOfInput => self.accept(index),
```
Comment (brendanzab) [inline src/main.rs:310]: Does this make sense?

File: src/main.rs [hunk 3]
```diff
@@ -71,6 +71,8 @@ enum Format {
     Fail,
     /// A format that matches the empty byte string
     Empty,
+    /// Matches if the end of the input has be reached
```
Comment (mikeday) [inline src/main.rs:74]: s/be/been/

File: src/main.rs (hunk 1 above)
Comment (mikeday) [reply src/main.rs:240]: I think `EndOfInput` is nullable? Given that it's equivalent to `Empty` with the added constraint that it must occur at the end of the input. So for example `Repeat(EndOfInput)` at the end of the file would present a problem!

Comment (mikeday) [reply src/main.rs:240]: Thinking about it some more yes it's definitely nullable as `EndOfInput` can match the empty string.

File: src/main.rs (hunk 2 above)
Comment (mikeday) [reply src/main.rs:310]: Yes I think this makes sense.

**Review by mikeday — APPROVED** [review]
Looks good! Would be nice to have a test for `Cat(Repeat(0x00), EndOfInput)`

File: src/main.rs (hunk 1 above)
Comment (brendanzab) [reply src/main.rs]: I added a doc comment that hopefully clarifies the purpose of this method some more

**Comment by brendanzab** [comment]
Test added, hopefully the method rename is ok too!

File: src/main.rs [hunk 4]
```diff
@@ -1532,6 +1579,20 @@ mod tests {
         );
     }
 
+    #[test]
+    fn compile_cat_end_of_input() {
+        let f = Format::Cat(Box::new(is_byte(0x00)), Box::new(Format::EndOfInput));
```
Comment (mikeday) [inline src/main.rs:1584]: Oh I was thinking this could use Repeat to check that compilation works and the loop knows when to stop

**Comment by mikeday** [comment]
Yep the rename is fine!

**Pull request merged**